autokeras/engine/tuner.py

```diff
@@ -47,46 +48,31 @@ def get_best_model(self):
         model.load_weights(self.best_model_path)
         return model
 
-    def _super_run_trial(self, trial, x=None, *fit_args, **fit_kwargs):
-        # TODO: Remove this function after TF can adapt in fit.
-        # Handle any callbacks passed to `fit`.
-        copied_fit_kwargs = copy.copy(fit_kwargs)
-        callbacks = fit_kwargs.pop('callbacks', [])
-        callbacks = self._deepcopy_callbacks(callbacks)
-        self._configure_tensorboard_dir(callbacks, trial.trial_id)
-        # `TunerCallback` calls:
-        # - `Tuner.on_epoch_begin`
-        # - `Tuner.on_batch_begin`
-        # - `Tuner.on_batch_end`
-        # - `Tuner.on_epoch_end`
-        # These methods report results to the `Oracle` and save the trained Model. If
-        # you are subclassing `Tuner` to write a custom training loop, you should
-        # make calls to these methods within `run_trial`.
-        callbacks.append(tuner_utils.TunerCallback(self, trial))
-        copied_fit_kwargs['callbacks'] = callbacks
-
-        model = self.hypermodel.build(trial.hyperparameters)
-        utils.adapt_model(model, x)
-        model.fit(x, *fit_args, **copied_fit_kwargs)
-
-    def run_trial(self,
-                  trial,
-                  x=None,
-                  validation_data=None,
-                  *fit_args,
-                  **fit_kwargs):
-        x, validation_data = self.build_preprocessors(
-            trial.hyperparameters, x, validation_data)
-        self._super_run_trial(
-            trial=trial,
-            x=x,
-            validation_data=validation_data,
-            *fit_args,
-            **fit_kwargs)
-
-    def build_preprocessors(self, hp, x=None, validation_data=None):
-        # TODO: Split x and build the preprocessor for each input node.
-        return x, validation_data
+    def _on_train_end(self, model, hp, x, *args, **kwargs):
+        self.adapt(model, x)
+
+    @staticmethod
+    def adapt(model, dataset):
```

Comment:
The logic in this method will only work for single-input, single-output stacks of layers (Sequential or Sequential-like models). Will it be necessary to support arbitrary topologies in the future?

Reply:
Yes, the logic now is quite limited, but good enough to support the basic use cases of AutoKeras now. We will leverage [PreprocessingStage](https://github.com/keras-team/governance/blob/master/rfcs/20190502-preprocessing-layers.md#prepocessingstage) to improve it in the future. Eventually, this may be moved to keras tuner for supporting preprocessing layers.